sbi/inference/trainers/base.py

```diff
@@ -873,6 +894,203 @@ def _create_posterior(
+                prevent exploding gradients. Use None for no clipping.
+            loss_kwargs: Additional or updated kwargs to be passed to the self._loss fn.
+
+        Returns:
+            The average training loss over all samples in the epoch.
+        """
+
+        assert self._neural_net is not None
+
+        train_loss_sum = 0
+        for batch in train_loader:
+            self.optimizer.zero_grad()
+            train_losses = self._get_training_losses(batch, loss_kwargs=loss_kwargs)
+            train_loss = torch.mean(train_losses)
+            train_loss_sum += train_losses.sum().item()
+
+            train_loss.backward()
+            if clip_max_norm is not None:
+                clip_grad_norm_(
+                    self._neural_net.parameters(),
+                    max_norm=clip_max_norm,
+                )
+            self.optimizer.step()
+
+        train_loss_average = train_loss_sum / (
+            len(train_loader) * train_loader.batch_size  # type: ignore
+        )
+
+        return train_loss_average
+
+    def _validate_epoch(
+        self,
+        val_loader: data.DataLoader,
+        loss_kwargs: Dict[str, Any],
+    ) -> float:
+        """
+        Perform a single validation epoch over the provided validation data.
+
+        Args:
+            val_loader: Dataloader for validation.
+            loss_kwargs: Additional or updated kwargs to be passed to the self._loss fn.
+
+        Returns:
+            The average validation loss over all samples in the epoch.
+        """
+
+        val_loss_sum = 0
+        with torch.no_grad():
+            for batch in val_loader:
+                val_losses = self._get_validation_losses(
+                    batch=batch,
+                    loss_kwargs=loss_kwargs,
+                )
+                val_loss_sum += val_losses.sum().item()
+
+        # Take mean over all validation samples.
+        val_loss = val_loss_sum / (
+            len(val_loader) * val_loader.batch_size  # type: ignore
+        )
+
+        return val_loss
+
+    def _summarize_epoch(
+        self,
+        train_loss: float,
+        val_loss: float,
+        epoch_start_time: float,
+        summarization_kwargs: Dict[str, Any],
+    ) -> None:
+        """
+        Update internal summaries after a single training epoch.
+
+        Records training and validation losses, as well as the duration of the epoch,
+        in `self._summary` dictionary.
+
+        Args:
+            train_loss: The average training loss for the epoch.
+            val_loss: The average validation loss for the epoch.
+            epoch_start_time: Timestamp when the epoch started, used to compute
+                duration.
+            summarization_kwargs: Additional keyword arguments for customizing
+                the summarization.
+        """
+
+        self._summary["training_loss"].append(train_loss)
+
+        self._val_loss = val_loss
```

Comment:
setting this `val_loss` goes beyond just "summarizing the epoch" as the function name suggests. is this then used somewhere else? 
maybe do it explicitly in `_run_training_loop`?